File: c++ code/week1/examples/basic functions/basic functions/warpaffine.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

#include <iostream>
#include <cmath>
#include <vector>

int main4()
{
	// read image
	cv::Mat source = cv::imread("../../../../../images/lena.jpg");

	// create 2 warp  matrices for diferent transformations
	cv::Mat warpMat1 = (cv::Mat_<double>(2, 3) << 1.2, 0.2, 2, -0.3, 1.3, 1);
	cv::Mat warpMat2 = (cv::Mat_<double>(2, 3) << 1.2, 0.3, 2, 0.2, 1.3, 1);

	cv::Mat result1, result2;

	// use warp affine
	cv::warpAffine(source, result1, warpMat1, cv::Size(source.cols, source.rows), cv::INTER_LINEAR, cv::BORDER_REFLECT_101);
	cv::warpAffine(source, result2, warpMat2, cv::Size(source.rows, source.cols), cv::INTER_LINEAR, cv::BORDER_REFLECT_101);

	// display image
	cv::imshow("Original", source);
	cv::imshow("Result1", result1);
	cv::imshow("Result2", result2);

	cv::waitKey(0);

	return EXIT_SUCCESS;
}

File: python code/basic functions/warpaffine.py
import cv2
import numpy as np

# read image
source = cv2.imread("../../images/lena.jpg", flags=1)

# create 2 warp matrices / masks for different transformations
warpMat1 = np.float32([[1.2, 0.2, 2], [-0.3, 1.3, 1]])
warpMat2 = np.float32([[1.2, 0.3, 2], [0.2, 1.3, 1]])

# use warp affine
dim = source.shape

result1 = cv2.warpAffine(source, warpMat1, (int(1.5 * dim[0]), int(1.4 * dim[1])),
                         None, flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_REFLECT_101)

result2 = cv2.warpAffine(source, warpMat2, (int(1.5 * dim[0]), int(1.4 * dim[1])),
                         None, flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_REFLECT_101)

# display images
cv2.imshow("Original", source)
cv2.imshow("Result1", result1)
cv2.imshow("Result2", result2)

cv2.waitKey(0)
